### PR TITLE
[stdlib] Tuple.__contains__

### PR DESCRIPTION
**Pull request by rd4com**
Hello,

here is an implementation of ```Tuple.__contains__```,

it makes it possible to do: 
```mojo
  var x = Tuple(1,2,True)
  if 1 in x: print("x contains 1")
```

@laszlokindrat suggested a ping for the PR in #2658

### CODE REVIEW

File: stdlib/test/builtin/test_tuple.mojo
Comment (laszlokindrat) [inline stdlib/test/builtin/test_tuple.mojo:19]: Can you please change this to not have `1` (or) `0` in it? Since Bool would convert to that implicitly, it can get tricky if something goes wrong with the call emission.

File: stdlib/test/builtin/test_tuple.mojo
Comment (laszlokindrat) [inline stdlib/test/builtin/test_tuple.mojo:22]: Can you please add equivalent tests that call into `__contains__` directly? I'm still paranoid about these features.

File: stdlib/test/builtin/test_tuple.mojo
Comment (laszlokindrat) [inline stdlib/test/builtin/test_tuple.mojo:23]: Can you please add an `assert_false` case with some different type, e.g. `assert_false("foo" in a)`?

**Review by laszlokindrat — APPROVED** [review]
Pretty cool, thank you!

**Comment by rd4com** [comment]
@laszlokindrat Yes, sorry, the tests are more complete now, do you see more cases to add ?

There is a test failing in test_list.mojo for macosx, not sure what to do ?

File: stdlib/src/builtin/tuple.mojo
Comment (soraros) [inline stdlib/src/builtin/tuple.mojo:210]: @laszlokindrat I don't think `rebind` will work for mem type.

Comment (rd4com) [reply stdlib/src/builtin/tuple.mojo]: Good observation! I changed the implementation to not use rebind and changed rhs to owned instead of borrowed.

Tests runs fine on ubuntu locally but somehow fails on macosx here, and stop the testing process of ubuntu?

Please see results on the (macos-14)  tab: 
```mojo
Please file a bug report.
[6505:28327:20240517,222150.691559:WARNING crash_report_exception_handler.cc:257] UniversalExceptionRaise: (os/kern) failure (5)
```

File: docs/changelog.md
Comment (soraros) [inline docs/changelog.md:281]: Nit: maybe reformat it? Same for the doc string.

File: stdlib/src/builtin/tuple.mojo
Comment (laszlokindrat) [reply stdlib/src/builtin/tuple.mojo]: Why does the rhs need to be owned?

Comment (rd4com) [reply stdlib/src/builtin/tuple.mojo]: In hope to not worry about borrowed references and manually extend lifetime with _=^,
Last commit passes tests, but there must be a better way to cast a reference than that ?

File: stdlib/src/builtin/tuple.mojo
Comment (laszlokindrat) [inline stdlib/src/builtin/tuple.mojo:186]: Please use PascalCase for type parameters.

File: stdlib/src/builtin/tuple.mojo
Comment (laszlokindrat) [reply stdlib/src/builtin/tuple.mojo]: > In hope to not worry about borrowed references and manually extend lifetime with _=^,

Unfortunately, `owned` can cause unintended copies, so I would rather avoid it.

File: stdlib/src/builtin/tuple.mojo
Comment (rd4com) [reply stdlib/src/builtin/tuple.mojo:186]: Yes, sorry about that, i do remember it all the time i see parameters now so thanks  :+1: 

Have got a better implementation, safer with references, please do not merge this one yet if it was the plan;

I'll do a force-push later-on, still don't know if it will work with the rebased!

File: stdlib/src/builtin/tuple.mojo
Comment (rd4com) [reply stdlib/src/builtin/tuple.mojo]: Hello, i forced-push a commit with changes related to both suggestions (no owned rhs and no rebind of T),
 i suggest rebinding the reference instead of T, since ```Reference``` is ```register_passable``` but not always T.
Extended test with memory-only types passes for ```Tuple.__contains__``` on last commit!
If both of you likes it, we could resolve this specific review conversation

**Comment by rd4com** [comment]
The force-push worked with the rebased branch :partying_face: :
- it includes a new ```@parameter``` optimization that returns false if ```T``` not in ```self.element_types```
  *(before ```var result:Bool = False```)*
- no rebind of memory-only types, we rebind the ```Reference```, which is ```@register_passable```

Todo:
- when ```@parameter for``` can do returns, we will be able to return True the first time that ```self[i] == value```,
instead of continuing doing result |= for every ```_type_is_eq[element_types[i],T]()```.
There would be no need for ```var result:Bool```

ℹ️ Note that the rebase is on ```2024.5.1805``` not  ```2024.5.1905```

**Comment by laszlokindrat** [comment]
!sync

**Review by laszlokindrat — APPROVED** [review]
Thanks again! I think this is a useful functionality, and while we can't do it perfectly yet, it does deliver a useful feature, especially for people coming from Python. There are very likely some sharp edges around the current behavior. If you can think of any, please open a PR to document them.

**Comment by rd4com** [comment]
@laszlokindrat , This is great! this is a cool feature, i'd be happy to use it too :smile: .
&nbsp;

Once auto-dereference lands:
- It will probably require some changes, as it might rebind the dereferenced value instead of the ```Reference```.

If there is any way you can add this to sort of a todo list to make sure ```Tuple.__contains__``` still works?

**Comment by laszlokindrat** [comment]
> @laszlokindrat , This is great! this is a cool feature, i'd be happy to use it too 😄 .  
> 
> Once auto-dereference lands:
> 
> * It will probably require some changes, as it might rebind the dereferenced value instead of the `Reference`.
> 
> If there is any way you can add this to sort of a todo list to make sure `Tuple.__contains__` still works?

The tests should prevent it from being broken.

**Pull request closed without merging**
